qiskit/algorithms/fidelities/base_fidelity.py

```diff
@@ -0,0 +1,253 @@
+            circuits_2: (Parametrized) quantum circuits.
+
+        Raises:
+            ValueError: if the length of the input circuit lists doesn't match.
+        """
+
+        if not len(circuits_1) == len(circuits_2):
+            raise ValueError(
+                f"The length of the first circuit list({len(circuits_1)}) "
+                f"and second circuit list ({len(circuits_2)}) does not coincide."
+            )
+
+        circuits = []
+        for (circuit_1, circuit_2) in zip(circuits_1, circuits_2):
+
+            circuit = self._circuit_cache.get((id(circuit_1), id(circuit_2)))
+
+            if circuit is not None:
+                circuits.append(circuit)
+            else:
+                self._check_qubits_match(circuit_1, circuit_2)
+
+                # re-parametrize input circuits
+                left_parameters = ParameterVector("x", circuit_1.num_parameters)
+                self._left_parameters.append(left_parameters)
+                parametrized_circuit_1 = circuit_1.assign_parameters(left_parameters)
+
+                right_parameters = ParameterVector("y", circuit_2.num_parameters)
+                self._right_parameters.append(right_parameters)
+                parametrized_circuit_2 = circuit_2.assign_parameters(right_parameters)
+
+                circuit = self._create_fidelity_circuit(
+                    parametrized_circuit_1, parametrized_circuit_2
+                )
+                circuits.append(circuit)
+                # update cache
+                self._circuit_cache[id(circuit_1), id(circuit_2)] = circuit
+
+        # set circuits
+        self._circuits = circuits
+
+    def _set_values(
+        self, values_1: Sequence[Sequence[float]] | None, values_2: Sequence[Sequence[float]] | None
+    ) -> None:
+        """
+        Update the list of parameter values to evaluate the corresponding
+        fidelity circuits with.
+
+        Args:
+            values_1: Numerical parameters to be bound to the first circuits.
+            values_2: Numerical parameters to be bound to the second circuits.
+
+        Raises:
+            ValueError: if the length of the input value lists doesn't match.
+        """
+
+        values = []
+        if values_2 is not None or values_1 is not None:
+            if values_2 is None:
+                values = values_1
+            elif values_1 is None:
+                values = values_2
+            else:
+                for (val_1, val_2) in zip(values_1, values_2):
+                    if len(val_1) != len(val_2):
+                        raise ValueError(
+                            f"The number of left parameters (currently {len(val_1)})"
+                            f"has to be equal to the number of right parameters."
+                            f"(currently {len(val_2)})"
+                        )
+                    values.append(val_1 + val_2)
+
+        # set values
+        self._parameter_values = values
+
+    def _preprocess_inputs(
```

Comment:
This is just a suggestion, so feel free to ignore 🙂 

Right now I have to know that I need to call this method in `_run` but it would not be quite clear to me that this validates the inputs and sets internal variables. Another solution could be to have two functions in the base class:
```python
def construct_circuit(self, circuit1, circuit2) -> QuantumCircuit:
     """Constructs the circuit to evaluate the fidelity."""
     
def validate_inputs(self, circuit1, circuit2, values1, values2) -> Tuple[np.array | None, np.array | None]:
     """Checks the inputs and returns the values in the right shape."""
```
and then the fidelity instances would look something like
```python
def _run(...):
     values1, values2 = self.validate_inputs(...)
     circuit = self.construct_circuit(....)
```
which might be more intuitive to use -- but that's just my personal opinion!

Reply:
It does make the `_run()` method more readable, so I applied a version of your suggestion, but used different method names. Thx!